packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,513 @@
+    if (_textDirection == value) {
+      return;
+    }
+    _textDirection = value;
+    markNeedsLayout();
+  }
+
+  List<Paint> get fillPaint => _fillPaint;
+  final List<Paint> _fillPaint = <Paint>[];
+
+  @override
+  double computeMinIntrinsicWidth(double height) => constraints.maxWidth / childCount;
+
+  @override
+  double computeMaxIntrinsicWidth(double height) => constraints.maxWidth / childCount;
+
+  @override
+  double computeMinIntrinsicHeight(double width) {
+    RenderBox child = firstChild;
+    double minHeight = double.infinity;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      if (child.computeMaxIntrinsicHeight(constraints.maxWidth / childCount) < minHeight) {
+        minHeight = child.computeMaxIntrinsicHeight(constraints.maxWidth / childCount);
+      }
+      child = childParentData.nextSibling;
+    }
+    return minHeight;
+  }
+
+  @override
+  double computeMaxIntrinsicHeight(double width) => computeMinIntrinsicHeight(width);
+
+  @override
+  double computeDistanceToActualBaseline(TextBaseline baseline) {
+    return defaultComputeDistanceToHighestActualBaseline(baseline);
+  }
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! _SegmentedControlContainerBoxParentData) {
+      child.parentData = new _SegmentedControlContainerBoxParentData();
+    }
+  }
+
+  void layoutRects(Function nextChild, RenderBox leftChild, RenderBox rightChild) {
+    RenderBox child = leftChild;
+    double start = 0.0;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      final Offset childOffset = Offset(start, 0.0);
+      childParentData.offset = childOffset;
+      final Rect childRect = Rect.fromLTWH(start, 0.0, child.size.width, child.size.height);
+      RRect rChildRect;
+      if (child == leftChild) {
+        rChildRect = RRect.fromRectAndCorners(childRect, topLeft: const Radius.circular(3.0),
+            bottomLeft: const Radius.circular(3.0));
+      } else if (child == rightChild) {
+        rChildRect = RRect.fromRectAndCorners(childRect, topRight: const Radius.circular(3.0),
+            bottomRight: const Radius.circular(3.0));
+      } else {
+        rChildRect = RRect.fromRectAndCorners(childRect);
+      }
+      childParentData.surroundingRect = rChildRect;
+      start += child.size.width;
+      child = nextChild();
+    }
+  }
+
+  @override
+  void performLayout() {
+    double maxHeight = _kMinSegmentedControlHeight;
+    final double childWidth = constraints.maxWidth / childCount;
+
+    RenderBox child = firstChild;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      final double boxHeight = child.getMaxIntrinsicHeight(childWidth);
+      if (boxHeight > maxHeight) {
+        maxHeight = boxHeight;
+      }
+      child = childParentData.nextSibling;
+    }
```

Comment:
make sure to constrain the maxHeight to the incoming `constraints` (using `constraints.constrainHeight`).